File: common/src/main/java/org/sandboxpowered/silica/util/SilicaIdentity.java
package org.sandboxpowered.silica.util;

import com.google.common.base.Objects;
import org.sandboxpowered.api.util.Identity;

public class SilicaIdentity implements Identity {
    public final String namespace;
    public final String path;

    public SilicaIdentity(String namespace, String path) {
        this.namespace = namespace;
        this.path = path;
    }

    public String getNamespace() {
        return namespace;
    }

    public String getPath() {
        return path;
    }

    @Override
    public String toString() {
        return namespace + ":" + path;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SilicaIdentity that = (SilicaIdentity) o;
        return Objects.equal(namespace, that.namespace) && Objects.equal(path, that.path);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(namespace, path);
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/util/SilicaIdentity.kt
package org.sandboxpowered.silica.util

import com.google.common.base.Objects
import org.sandboxpowered.api.util.Identifier

class SilicaIdentity(private val namespace: String, private val path: String) : Identifier {
    override fun getNamespace(): String {
        return namespace
    }

    override fun getPath(): String {
        return path
    }

    override fun toString(): String {
        return "$namespace:$path"
    }

    override fun compareTo(other: Identifier): Int {
        return when (val ns = namespace.compareTo(other.namespace)) {
            0 -> path.compareTo(other.path)
            else -> ns
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as SilicaIdentity
        return Objects.equal(namespace, that.namespace) && Objects.equal(
            path, that.path
        )
    }

    override fun hashCode(): Int {
        return Objects.hashCode(namespace, path)
    }
}